### PR TITLE
UsingRegistryV2 contract with static registry address.

### LINKED ISSUE #7616: [Bounty] FeeCurrencyWhitelist is not UsingRegistry storage variable
### Expected Behavior

Because the `Registry` address `0x0..ce10` (and for that matter the `SortedOracles` address) are not expected to ever change, this contract can make the address static in its bytecode and avoid a complex `Governance` situation. 

In future it could save this and other core contracts a marginal amount of gas avoiding the relevant storage load operations. 

This decision and migration from a storage registry address to a bytecode registry address should be well documented and understood.

### Current Behavior

In this [gitcoin bounty PR](https://github.com/celo-org/celo-monorepo/pull/7550), it was discovered that `FeeCurrencyWhitelist` is not already consuming `UsingRegistry` for lack of use. 

However, the [ticket](https://github.com/celo-org/celo-monorepo/issues/7315) specifies that fee currencies should have oracle reports before they are whitelisted, meaning a lookup to the `SortedOracles` contract through the registry. Introducing `FeeCurrencyWhitelist is UsingRegistry` implies a storage level change which requires a new proxy deployment and initialization, and an atomic migration from the existing whitelist.

### PR DESCRIPTION
**Pull request by deepakhb2**
### Description

Add UsingRegistryV2 contract and make registry address static in its bytecode.
Add Dollar and Euro identifier get methods for stable token and exchange.

### Other changes

NIL

### Tested

Doesn't require unit tests as it is always extended by other contracts.

### Related issues

- Fixes #7616 

### Backwards compatibility

New contract.

### Documentation

TODO

### CODE REVIEW

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+import "../identity/interfaces/IRandom.sol";
+import "../identity/interfaces/IAttestations.sol";
+
+import "../stability/interfaces/IExchange.sol";
+import "../stability/interfaces/IReserve.sol";
+import "../stability/interfaces/ISortedOracles.sol";
+import "../stability/interfaces/IStableToken.sol";
+
+contract UsingRegistryV2 is Ownable {
```
Comment (yorhodes) [inline packages/protocol/contracts/common/UsingRegistryV2.sol]: We shouldn't require that consumers of `UsingRegistryV2` are also `Ownable`
```suggestion
contract UsingRegistryV2 {
```

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  address constant registryAddress = 0x000000000000000000000000000000000000ce10;
+  IRegistry public registry = IRegistry(registryAddress);
```
Comment (yorhodes) [inline packages/protocol/contracts/common/UsingRegistryV2.sol]: I think `registry` can be `constant` too?

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    return IReserve(registry.getAddressForOrDie(RESERVE_REGISTRY_ID));
+  }
+
+  function getSortedOracles() internal view returns (ISortedOracles) {
+    return ISortedOracles(registry.getAddressForOrDie(SORTED_ORACLES_REGISTRY_ID));
+  }
+
+  function getStableToken() internal view returns (IStableToken) {
+    return IStableToken(registry.getAddressForOrDie(STABLE_TOKEN_REGISTRY_ID));
+  }
```
Comment (yorhodes) [inline packages/protocol/contracts/common/UsingRegistryV2.sol:124]: we now have a celo euro which should also be included here
I don't think we should deprecate the `getStableToken` but maybe we can provide an additional `getStableDollarToken` and `getStableEuroToken`

Comment (yorhodes) [reply packages/protocol/contracts/common/UsingRegistryV2.sol:124]: the same goes for `Exchange`

**Review by yorhodes — COMMENTED** [review]
great start!

File: packages/protocol/contracts/common/UsingRegistryV2.sol (hunk 3 above)
Comment (deepakhb2) [reply packages/protocol/contracts/common/UsingRegistryV2.sol:124]: @yorhodes you mean we need to add get methods for Dollar and Euro? ex. getDollarExchange.
Also, should I need to create different interfaces that extend form IStableToken for Dollar and Euro?

Comment (yorhodes) [reply packages/protocol/contracts/common/UsingRegistryV2.sol:124]: its the same interfave IStableToken and IExchange
just different identifiers

**Comment by deepakhb2** [comment]
@yorhodes Please let me know how this change needs to be documented. I see you have mentioned about documentation in the issue.

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  bytes32 constant ACCOUNTS_REGISTRY_ID = keccak256(abi.encodePacked("Accounts"));
+  bytes32 constant ATTESTATIONS_REGISTRY_ID = keccak256(abi.encodePacked("Attestations"));
+  bytes32 constant DOWNTIME_SLASHER_REGISTRY_ID = keccak256(abi.encodePacked("DowntimeSlasher"));
+  bytes32 constant DOUBLE_SIGNING_SLASHER_REGISTRY_ID = keccak256(
+    abi.encodePacked("DoubleSigningSlasher")
+  );
+  bytes32 constant ELECTION_REGISTRY_ID = keccak256(abi.encodePacked("Election"));
+  bytes32 constant EXCHANGE_REGISTRY_ID = keccak256(abi.encodePacked("Exchange"));
+  bytes32 constant EXCHANGE_DOLLAR_REGISTRY_ID = keccak256(abi.encodePacked("ExchangeDollar"));
+  bytes32 constant EXCHANGE_EURO_REGISTRY_ID = keccak256(abi.encodePacked("ExchangeEuro"));
```
Comment (yorhodes) [inline packages/protocol/contracts/common/UsingRegistryV2.sol]:
```suggestion
  bytes32 constant EXCHANGE_REGISTRY_ID = keccak256(abi.encodePacked("Exchange"));
  bytes32 constant EXCHANGE_EURO_REGISTRY_ID = keccak256(abi.encodePacked("ExchangeEUR"));
```

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  );
+  bytes32 constant LOCKED_GOLD_REGISTRY_ID = keccak256(abi.encodePacked("LockedGold"));
+  bytes32 constant RESERVE_REGISTRY_ID = keccak256(abi.encodePacked("Reserve"));
+  bytes32 constant RANDOM_REGISTRY_ID = keccak256(abi.encodePacked("Random"));
+  bytes32 constant SORTED_ORACLES_REGISTRY_ID = keccak256(abi.encodePacked("SortedOracles"));
+  bytes32 constant STABLE_TOKEN_REGISTRY_ID = keccak256(abi.encodePacked("StableToken"));
+  bytes32 constant STABLE_DOLLOR_TOKEN_REGISTRY_ID = keccak256(
+    abi.encodePacked("StableDollarToken")
+  );
+  bytes32 constant STABLE_EURO_TOKEN_REGISTRY_ID = keccak256(abi.encodePacked("StableEuroToken"));
```
Comment (yorhodes) [inline packages/protocol/contracts/common/UsingRegistryV2.sol]:
```suggestion
  bytes32 constant STABLE_TOKEN_REGISTRY_ID = keccak256(abi.encodePacked("StableToken"));
  bytes32 constant STABLE_EURO_TOKEN_REGISTRY_ID = keccak256(abi.encodePacked("StableTokenEUR"));
```

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  function getElection() internal view returns (IElection) {
+    return IElection(registry.getAddressForOrDie(ELECTION_REGISTRY_ID));
+  }
+
+  function getExchange() internal view returns (IExchange) {
+    return IExchange(registry.getAddressForOrDie(EXCHANGE_REGISTRY_ID));
+  }
+
+  function getExchangeDollar() internal view returns (IExchange) {
+    return IExchange(registry.getAddressForOrDie(EXCHANGE_DOLLAR_REGISTRY_ID));
```
Comment (yorhodes) [inline packages/protocol/contracts/common/UsingRegistryV2.sol]:
```suggestion
    return IExchange(registry.getAddressForOrDie(EXCHANGE_REGISTRY_ID));
  }

  function getExchangeDollar() internal view returns (IExchange) {
    return getExchange();
```

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  function getSortedOracles() internal view returns (ISortedOracles) {
+    return ISortedOracles(registry.getAddressForOrDie(SORTED_ORACLES_REGISTRY_ID));
+  }
+
+  function getStableToken() internal view returns (IStableToken) {
+    return IStableToken(registry.getAddressForOrDie(STABLE_TOKEN_REGISTRY_ID));
+  }
+
+  function getStableDollarToken() internal view returns (IStableToken) {
+    return IStableToken(registry.getAddressForOrDie(STABLE_DOLLOR_TOKEN_REGISTRY_ID));
```
Comment (yorhodes) [inline packages/protocol/contracts/common/UsingRegistryV2.sol]:
```suggestion
    return IStableToken(registry.getAddressForOrDie(STABLE_TOKEN_REGISTRY_ID));
  }

  function getStableDollarToken() internal view returns (IStableToken) {
    return getStableToken();
```

**Review by yorhodes — COMMENTED** [review]
some adjustments to canonical registry ids

**Review by yorhodes — APPROVED** [review]
Great!

**Pull request merged**